sentry-core/src/main/java/io/sentry/core/util/CollectionUtils.java

```diff
@@ -25,4 +28,28 @@ public static int size(Iterable<?> data) {
     }
     return counter;
   }
+
+  /**
+   * Creates a shallow copy of map given by parameter.
+   *
+   * @param map the map to copy
+   * @param <K> the type of map keys
+   * @param <V> the type of map values
+   * @return the shallow copy of map
+   */
+  public static <K, V> @Nullable Map<K, V> shallowCopy(@Nullable Map<K, V> map) {
+    if (map != null) {
+      final Map<K, V> clone = new HashMap<>();
```

Comment:
If we're cloning into non-thread-safe collections, couldn't we be cloning them later and end up crashing this for them being changed while we clone?

Reply:
I'd do something like:

`final Map<String, Object> unknownRef = unknown;`

check if `unknownRef` is null first and only use `unknownRef`

Reply:
After thinking it through, `HashMap` constructor does pretty much the same and if we change implementation to `ConcurrentHashMap` we would avoid `ConcurrentModificationException` if map gets modified while other thread iterates over it.

Whole method could be replaced with:

```java
public static <K, V> @Nullable Map<K, V> shallowCopy(@Nullable Map<K, V> map) {
  if (map != null) {
    return new ConcurrentHashMap<>(map);
  } else {
    return null;
  }
}
```

Reply:
I think the biggest problem is if `map` is set to null in a different thread, then it'd cause NPE I guess?!
so creating an aux. field reference for `map` might solve the problem.

Reply:
The only risk I see is that `map` can be modified while `ConcurrentHashMap` is being constructed. Perhaps `shallowCopy` should only accept `ConcurrentHashMap` as a parameter and all `unknown` and other maps in objects that are meant to put into `Contexts` should be also enforced to be a type of `ConcurrentHashMap`.

Reply:
making everything of the type `ConcurrentHashMap` has some overhead.
why not just:

```
final Map mapRef = map;
if (mapRef != null) {
    return new ConcurrentHashMap<>(mapRef);
} else {
    return null;
}
```

I believe only this would be enough

Reply:
`map` cannot be set to `null` in other thread as it's a copy of the reference of the original reference passed to this function, right?

`ConcurrentHashMap` in the constructor iterates over `map` passed as a parameter, meaning that if during this time this map gets modified and it's a plain `HashMap` I'm afraid `ConcurrentModificationException` may be thrown.

Reply:
Fix for that has been implemented in https://github.com/maciejwalkowiak/sentry-android/pull/1 (separate PR so that we can isolated thread-safety discussions).